backend/transcribee_backend/routers/task.py

```diff
@@ -64,67 +78,101 @@ def claim_unassigned_task(
-                col(Task.last_keepalive)
-                < now_tz_aware() - datetime.timedelta(seconds=settings.worker_timeout),
+                is_(Task.current_attempt_id, None),
+                Task.current_attempt.has(
+                    TaskAttempt.last_keepalive
+                    < now_tz_aware()
+                    - datetime.timedelta(seconds=settings.worker_timeout)
+                ),
             ),
-            is_(Task.is_completed, False),
+            ~(col(Task.state).in_([TaskState.COMPLETED, TaskState.FAILED])),
             ~blocking_tasks_exist,
         )
         .with_for_update()
     )
-    task = session.exec(statement).first()
-    if task is None:
-        return task
-    task.assigned_worker = authorized_worker
-    task.last_keepalive = now_tz_aware()
-    task.assigned_at = now_tz_aware()
+    return session.exec(statement).first()
+
+
+@task_router.post("/claim_unassigned_task/")
+def claim_unassigned_task(
+    session: Session = Depends(get_session),
+    authorized_worker: Worker = Depends(get_authorized_worker),
+    task_type: List[TaskType] = Query(),
+    now: datetime.datetime = Depends(now_tz_aware),
+) -> Optional[AssignedTaskResponse]:
+    while True:
+        task = get_ready_task(session, task_type)
+        if task is None:
+            session.commit()  # We might have marked tasks as failed
+            return
+        elif task.remaining_attempts <= 0:
+            mark_task_as_failed(session, task, now)
+        else:
+            break
+
+    finish_current_attempt(session, task, now=now)
+    attempt = TaskAttempt(
+        task=task, started_at=now, assigned_worker=authorized_worker, last_keepalive=now
+    )
+    session.add(attempt)
+    task.current_attempt = attempt
+    task.remaining_attempts -= 1
+    task.state = TaskState.ASSIGNED
+    task.state_changed_at = now
     session.add(task)
     session.commit()
     return AssignedTaskResponse.from_orm(task)
 
 
 @task_router.post("/{task_id}/keepalive/")
 def keepalive(
-    task_id: uuid.UUID,
     keepalive_data: KeepaliveBody = Body(),
     session: Session = Depends(get_session),
-    authorized_worker: Worker = Depends(get_authorized_worker),
+    task: Task = Depends(get_authorized_task),
 ) -> Optional[AssignedTaskResponse]:
-    statement = select(Task).where(Task.id == task_id)
-    task = session.exec(statement).one_or_none()
-    if task is None:
-        raise HTTPException(status_code=404)
-
-    if task.assigned_worker != authorized_worker:
-        raise HTTPException(status_code=403)
-
-    task.last_keepalive = now_tz_aware()
+    task.current_attempt.last_keepalive = now_tz_aware()
     if keepalive_data.progress:
-        task.progress = keepalive_data.progress
+        task.current_attempt.progress = keepalive_data.progress
+        session.add(task.current_attempt)
     session.add(task)
     session.commit()
     return AssignedTaskResponse.from_orm(task)
```

Comment:
Do we actually ever do something if the keepalive does not come on our own in the backend? (or only if a worker polls us?)